// ngx-french-toast/esm2022/lib/french-toast.module.mjs
import { NgModule } from '@angular/core';
import { ToastPosition } from './enums/enums';
import { TOAST_CONFIG } from './toast.tokens';
import * as i0 from "@angular/core";
export class FrenchToastModule {
    static forRoot(config) {
        if (!config) {
            config = {
                position: ToastPosition.BOTTOM_RIGHT,
                defaultDuration: 10000,
                colors: {
                    autoGradient: false
                }
            };
        }
        return {
            ngModule: FrenchToastModule,
            providers: [
                { provide: TOAST_CONFIG, useValue: config }
            ]
        };
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: FrenchToastModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.8", ngImport: i0, type: FrenchToastModule });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: FrenchToastModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: FrenchToastModule, decorators: [{
            type: NgModule,
            args: [{}]
        }] });
