// ngx-french-toast/esm2022/lib/french-toast.service.mjs
import { Inject, Injectable } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { ToastType } from './enums/enums';
import { TOAST_CONFIG } from './toast.tokens';
import { ComponentPortal } from '@angular/cdk/portal';
import { ToastsComponent } from './components/toasts/toasts.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
export class ToastService {
    config;
    overlay;
    toast = new BehaviorSubject(null);
    clearAll = new Subject();
    clearToast = new Subject();
    overlayRef;
    duration = 7000;
    constructor(config, overlay) {
        this.config = config;
        this.overlay = overlay;
        if (this.config?.defaultDuration) {
            this.duration = this.config.defaultDuration;
        }
    }
    addToast(toastInput, type) {
        const newToast = {
            _id: toastInput._id,
            title: toastInput.title,
            content: toastInput.content,
            isVisible: true,
            duration: toastInput.duration,
            icon: toastInput?.icon ?? null,
            type,
            component: toastInput.component,
            infinite: toastInput.infinite,
            pinned: toastInput.pinned,
            context: toastInput.context,
            _uId: this.getUniqueId(6)
        };
        this.toast.next(newToast);
        if (!this.overlayRef?.hasAttached())
            this.createOverlay();
    }
    createOverlay() {
        this.overlayRef = this.overlay.create();
        const toastPortal = new ComponentPortal(ToastsComponent);
        const componentRef = this.overlayRef.attach(toastPortal);
        componentRef.instance.componentRef = componentRef;
    }
    handleToast(toastInput, type) {
        toastInput._id = toastInput?._id ?? this.getUniqueId(5);
        toastInput.duration = toastInput?.duration ?? this.duration;
        this.addToast(toastInput, type);
    }
    success(toastInput) {
        this.handleToast(toastInput, ToastType.SUCCESS);
    }
    danger(toastInput) {
        this.handleToast(toastInput, ToastType.DANGER);
    }
    info(toastInput) {
        this.handleToast(toastInput, ToastType.INFO);
    }
    warning(toastInput) {
        this.handleToast(toastInput, ToastType.WARNING);
    }
    clearAllToasts() {
        this.clearAll.next();
    }
    destroyToast(toastComponent) {
        this.clearToast.next(toastComponent.toast._uId);
    }
    getUniqueId(parts) {
        const stringArr = [];
        for (let i = 0; i < parts; i++) {
            // tslint:disable-next-line:no-bitwise
            const S4 = (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
            stringArr.push(S4);
        }
        return stringArr.join('');
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: ToastService, deps: [{ token: TOAST_CONFIG }, { token: i1.Overlay }], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: ToastService, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.8", ngImport: i0, type: ToastService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TOAST_CONFIG]
                }] }, { type: i1.Overlay }] });
